### PR TITLE
Add documentation for `Rule.use` as a function

### LINKED ISSUE #2691: Rule.use as a function is not documented
`Rule.use` configuration is documented when used with a list of `UseEntry`, but not when used with a function.

I'll probably open a PR with something :smiley:

### PR DESCRIPTION
**Pull request by bew**
Adds some documentation for using a function for `Rule.use` configuration.

For the object passed to the function, I'm not surewhat `realResource`, `compiler` & `resourceQuery` are for, and I'd appreciate some help to explain them succinctly , thank you

Fixes #2691

### CODE REVIEW

File: src/content/configuration/module.md [hunk 1]
```diff
@@ -308,12 +308,27 @@ module.exports = {
 
 ## `Rule.use`
 
-A list of [UseEntries](#useentry) which are applied to modules. Each entry specifies a loader to be used.
+`array`
```
Comment (EugeneHlushko) [inline src/content/configuration/module.md]:
```suggestion
`[UseEntry]` `function`
```

Comment (EugeneHlushko) [reply src/content/configuration/module.md]: to be consistent with the rest of the docs

File: src/content/configuration/module.md [hunk 2]
```diff
@@ -308,12 +308,27 @@ module.exports = {
...
+
+An array of [UseEntry](#useentry) which are applied to modules. Each entry specifies a loader to be used.
 
 Passing a string (i.e. `use: [ 'style-loader' ]`) is a shortcut to the loader property (i.e. `use: [ { loader: 'style-loader '} ]`).
 
 Loaders can be chained by passing multiple loaders, which will be applied from right to left (last to first configured).
 
+`Function`
```
Comment (EugeneHlushko) [inline src/content/configuration/module.md]:
```suggestion
```

File: src/content/configuration/module.md [hunk 3]
```diff
@@ -308,12 +308,27 @@ module.exports = {
 
 ## `Rule.use`
 
-A list of [UseEntries](#useentry) which are applied to modules. Each entry specifies a loader to be used.
+`array`
+
+An array of [UseEntry](#useentry) which are applied to modules. Each entry specifies a loader to be used.
```
Comment (EugeneHlushko) [inline src/content/configuration/module.md]:
```suggestion
`Rule.use` can be an array of [UseEntry](#useentry) which are applied to modules. Each entry specifies a loader to be used.
```

File: src/content/configuration/module.md [hunk 4]
```diff
@@ -308,12 +308,27 @@ module.exports = {
...
 
 Passing a string (i.e. `use: [ 'style-loader' ]`) is a shortcut to the loader property (i.e. `use: [ { loader: 'style-loader '} ]`).
 
 Loaders can be chained by passing multiple loaders, which will be applied from right to left (last to first configured).
 
+`Function`
+
+A function which takes an object describing the module being loaded, and must return an array of `UseEntry`.
```
Comment (EugeneHlushko) [inline src/content/configuration/module.md]:
```suggestion
`Rule.use` can be a function which receives the object argument describing the module being loaded, and must return an array of `UseEntry`.
```

Although i am not sure it must return an array or single UseEntry

File: src/content/configuration/module.md [hunk 5]
```diff
@@ -308,12 +308,27 @@ module.exports = {
...
+
+The same shortcut as as an array can be used for the return value (i.e. `use: [ 'style-loader' ]`).
```
Comment (EugeneHlushko) [inline src/content/configuration/module.md]:
```suggestion
The same shortcut as an array can be used for the return value (i.e. `use: [ 'style-loader' ]`).
```

File: src/content/configuration/module.md [hunk 6]
```diff
@@ -308,12 +308,27 @@ module.exports = {
...
+
+The object parameter has the following fields:
```
Comment (EugeneHlushko) [inline src/content/configuration/module.md]: Move this info above the previous statement about shortcut. Please investigate question marks, wouldn't want to merge todos

**Comment by montogeek** [comment]
@bew Friendly ping

**Comment by bew** [comment]
I'm still not sure what `resourceQuery` is, it's never populated in the webpack config we use, anyone can help? Thank you

**Comment by EugeneHlushko** [comment]
The build fails, markdown lint doesnt pass with error:
```
./src/content/configuration/module.md: 324: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "- `resource`: The path to the ..."]
```

**Comment by EugeneHlushko** [comment]
`ResourceQuery` needs a look, dont have time today. If you dont have time either i would suggest  a change

File: src/content/configuration/module.md [hunk 7]
```diff
@@ -308,12 +308,25 @@ module.exports = {
...
 Loaders can be chained by passing multiple loaders, which will be applied from right to left (last to first configured).
 
+`Rule.use` can be a function which receives the object argument describing the module being loaded, and must return an array of `UseEntry`.
+
+The same shortcut as an array can be used for the return value (i.e. `use: [ 'style-loader' ]`).
+
+The object parameter has the following fields:
+- `resource`: The path to the module being loaded
+- `realResource`: The path to the module being loaded
+- `resourceQuery`: ??
```
Comment (EugeneHlushko) [inline src/content/configuration/module.md]:
```suggestion
 - `resourceQuery`
```

Comment (EugeneHlushko) [reply src/content/configuration/module.md]: Or investigate what it is

File: src/content/configuration/module.md [hunk 8]
```diff
@@ -308,12 +308,26 @@ module.exports = {
...
 Loaders can be chained by passing multiple loaders, which will be applied from right to left (last to first configured).
 
+`Rule.use` can be a function which receives the object argument describing the module being loaded, and must return an array of `UseEntry`.
+
+The object parameter has the following fields:
+
+- `compiler`: The current webpack compiler (can be undefined)
+- `issuer`: The path to the module that is importing the module being loaded
+- `resource`: The path to the module being loaded
+- `realResource`: The path to the module being loaded
```
Comment (EugeneHlushko) [inline src/content/configuration/module.md]: It is worth to note what differs `resource` and `realResource` here as it may confuse the reader @bew

Comment (bew) [reply src/content/configuration/module.md]: Indeed that would be better, but as stated in the comments above, I don't actually know what is the difference ^^
If you can help about this I'd be very grateful :)

Comment (sokra) [reply src/content/configuration/module.md]: `resource` is usually equal `realResource` expect when the `resource` name is overwritten via `!=!` in request string. `realResource` is always the resource file which is requested.

`require("LLL!RRR")` -> `resource: RRR, realResource: RRR`
`require("MMM!=!LLL!RRR")` -> `resource: MMM, realResource: RRR`

Comment (bew) [reply src/content/configuration/module.md]: Thanks, I'll add a note when I'm at work

**Comment by EugeneHlushko** [comment]
Thanks!

**Pull request merged**